File: src/main/java/szewek/flux/tile/DiggerTile.java
package szewek.flux.tile;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.loot.LootContext;
import net.minecraft.loot.LootParameters;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.vector.Vector3d;
import net.minecraft.world.server.ServerWorld;
import szewek.fl.util.ItemsUtil;
import szewek.fl.util.spatial.NonStopWalker;
import szewek.fl.util.spatial.WalkAction;
import szewek.flux.F;
import szewek.flux.FluxCfg;
import szewek.flux.block.ActiveTileBlock;

import java.util.List;

public final class DiggerTile extends BlockInteractingTile {
	private boolean lastFlag;

	public DiggerTile() {
		super(F.T.DIGGER, new NonStopWalker(-5, -256, -5, 5, -1, 5), FluxCfg.ENERGY.digger);
		walker.startFrom(true, false, true);
		walker.putActions(WalkAction.X_POS, WalkAction.Z_POS, WalkAction.Y_NEG);
	}

	@Override
	public void tick() {
		assert level != null;
		if (!level.isClientSide) {
			boolean flag = !disabled;
			final int usage = energyUse.get();
			if (flag && energy.getEnergyStored() >= usage) {
				walker.walk();
				BlockPos bp = walker.getPosOffset(worldPosition);

				if (bp.getY() < 0) {
					disabled = true;
					level.setBlock(worldPosition, level.getBlockState(worldPosition).setValue(ActiveTileBlock.LIT, false), 3);
					setChanged();
					return;
				}

				dig(bp);
				energy.use(usage);
			}

			if (flag != lastFlag) {
				level.setBlock(worldPosition, level.getBlockState(worldPosition).setValue(ActiveTileBlock.LIT, flag), 3);
				lastFlag = flag;
			}
		}

	}

	private void dig(BlockPos bp) {
		BlockState bs = level.getBlockState(bp);
		Block b = bs.getBlock();
		if (F.Tags.DIGGER_SKIP.contains(b) || b.hasTileEntity(bs)) {
			return;
		}
		List<ItemStack> drops = bs.getDrops(new LootContext.Builder((ServerWorld)level).withParameter(LootParameters.ORIGIN, Vector3d.atCenterOf(worldPosition)).withParameter(LootParameters.TOOL, ItemStack.EMPTY));
		if (!drops.isEmpty()) {
			level.removeBlock(bp, false);
			ItemsUtil.trySendingItems(drops, level, worldPosition);
		}
	}
}

File: src/main/java/szewek/flux/tile/DiggerTile.kt
package szewek.flux.tile

import net.minecraft.inventory.IInventory
import net.minecraft.inventory.InventoryHelper
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundNBT
import net.minecraft.util.Direction
import net.minecraft.world.server.ServerWorld
import net.minecraft.world.storage.loot.LootContext
import net.minecraft.world.storage.loot.LootParameters
import net.minecraftforge.common.Tags
import net.minecraftforge.items.CapabilityItemHandler
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.wrapper.InvWrapper
import szewek.flux.FTiles
import szewek.flux.block.DiggerBlock
import szewek.ktutils.getNeighborTile

class DiggerTile : PoweredTile(FTiles.DIGGER) {
    private var offsetX = 0
    private var offsetY = 0
    private var offsetZ = 0
    private var finished = false
    private var lastFlag = false

    override fun read(compound: CompoundNBT) {
        super.read(compound)
        offsetX = compound.getInt("OffX")
        offsetY = compound.getInt("OffY")
        offsetZ = compound.getInt("OffZ")
        finished = compound.getBoolean("Finished")
    }

    override fun write(compound: CompoundNBT): CompoundNBT {
        super.write(compound)
        compound.putInt("OffX", offsetX)
        compound.putInt("OffY", offsetY)
        compound.putInt("OffZ", offsetZ)
        compound.putBoolean("Finished", finished)
        return compound
    }

    override fun tick() {
        if (!world!!.isRemote) {
            val flag = energy >= 200 && !finished
            if (flag) {
                if (offsetY == 0 || (offsetX == 5 && offsetZ == 5)) {
                    offsetX = -5
                    offsetZ = -5
                    offsetY--
                } else if (offsetX == 5) {
                    offsetX = -5
                    offsetZ++
                } else {
                    offsetX++
                }
                val bp = pos.add(offsetX, offsetY, offsetZ)
                if (bp.y < 0) {
                    finished = true
                    return
                }
                val bs = world!!.getBlockState(bp)
                if (bs.block !in Tags.Blocks.DIRT && bs.block !in Tags.Blocks.STONE && bs.block !in Tags.Blocks.COBBLESTONE) {
                    val drops = bs.getDrops(LootContext.Builder(world as ServerWorld).withParameter(LootParameters.POSITION, pos).withParameter(LootParameters.TOOL, ItemStack.EMPTY))
                    val inv = mutableListOf<IItemHandler>()
                    if (drops.isNotEmpty()) {
                        world!!.removeBlock(bp, false)
                        for (dir in Direction.values()) {
                            if (dir == Direction.DOWN) continue
                            val te = getNeighborTile(dir)
                            if (te != null) {
                                lateinit var iih: IItemHandler
                                val il = te.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, dir.opposite)
                                if (il.isPresent) il.ifPresent {
                                    iih = it
                                } else if (te is IInventory) {
                                    iih = InvWrapper(te)
                                } else continue
                                inv += iih
                            }
                        }

                        itemloop@ for (id in drops) {
                            var stack = id
                            for (iih in inv) {
                                val l = iih.slots
                                for (i in 0 until l) {
                                    if (iih.isItemValid(i, stack)) {
                                        stack = iih.insertItem(i, stack, false)
                                    }
                                }
                                if (stack.isEmpty) break@itemloop
                            }
                            InventoryHelper.spawnItemStack(world!!, pos.x.toDouble(), (pos.y + 1).toDouble(), pos.z.toDouble(), stack)
                        }
                    }
                }
                energy -= 200
            }
            if (flag != lastFlag) {
                world!!.setBlockState(pos, world!!.getBlockState(pos).with(DiggerBlock.LIT, flag), 3)
                lastFlag = flag
            }
        }
    }
}